### PR TITLE
Sc 48 crud

### PR DESCRIPTION
**Pull request by ArezouRa**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
CRUD operations and tests for messages and Communication

## Related Tickets & Documents

- Related Issue #
- Closes #

## Added/updated tests?

_It is encouraged to keep the code coverage percentage at 80% and above._

- [X] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: src/messaging/views.py [hunk 1]
```diff
@@ -1,3 +1,51 @@
...
+class CommunicationList(generics.ListCreateAPIView):
+    """This class handles listing and creating communications in our REST API."""
+
+    queryset = Communication.objects.all()
+    serializer_class = CommunicationSerializer
+
+    def perform_create(self, serializer):
+        # Set the user field to the current user
+        serializer.save(user=self.request.user)
+
```
Comment (nomisreual) [inline src/messaging/views.py:39]: only fetch communications of current user

File: src/messaging/views.py [hunk 2]
```diff
@@ -1,3 +1,51 @@
...
+    """This class handles operations for a single communication instance.
+
+    - GET: Retrieve the details of a specific communication.
+    - PUT: Update the entire communication instance.
+    - PATCH: Partially update the communication instance.
+    - DELETE: Remove the specific communication instance.
+    """
+
+    queryset = Communication.objects.all()
+    serializer_class = CommunicationSerializer
```
Comment (nomisreual) [inline src/messaging/views.py:51]: again only for currently authenticated user

**Review by nomisreual — CHANGES_REQUESTED** [review]
One missing field in models.

Messages and communication access should be restricted.

File: src/messaging/views.py [hunk 3]
```diff
@@ -1,3 +1,51 @@
...
+from messaging.serializers import MessageSerializer, CommunicationSerializer
+from rest_framework import generics
 
-# Create your views here.
+
+class MessageList(generics.ListCreateAPIView):
+    """This class handles listing and creating messages in our REST API."""
+
+    queryset = Message.objects.all()
+    serializer_class = MessageSerializer
```
[the remark this answers is not in the document]
Comment (ArezouRa) [reply src/messaging/views.py:14]: tried that. would be glad to have your review

File: src/messaging/models.py [hunk 4]
```diff
@@ -1,3 +1,57 @@
...
+        """
+        return self.status == "accepted"
+
+
+class Message(models.Model):
+    """The Message model represents a message sent within a communication."""
+
+    communication = models.ForeignKey(Communication, on_delete=models.CASCADE)
+    created_at = models.DateTimeField(auto_now_add=True)
+    msg = models.TextField()
```
[the remark this answers is not in the document]
Comment (ArezouRa) [reply src/messaging/models.py:73]: is added and is waiting for review

File: src/messaging/views.py [hunk 5]
```diff
@@ -1,3 +1,51 @@
...
+    """This class handles operations for a single message instance.
+
+    - GET: Retrieve the details of a specific message.
+    - PUT: Update the entire message instance.
+    - PATCH: Partially update the message instance.
+    - DELETE: Remove the specific message instance.
+    """
+
+    queryset = Message.objects.all()
+    serializer_class = MessageSerializer
```
[the remark this answers is not in the document]
Comment (ArezouRa) [reply src/messaging/views.py:49]: tried to implement. hope that's what you wanted

File: src/messaging/serializers.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
+from rest_framework import serializers
+from messaging.models import Message, Communication
+
+
+class MessageSerializer(serializers.ModelSerializer):
+    # Use a PrimaryKeyRelatedField for communication to reference it by ID
+    communication = serializers.PrimaryKeyRelatedField(
+        queryset=Communication.objects.all()
+    )
```
Comment (nomisreual) [inline src/messaging/serializers.py:9]: so this grabs all messages authored by the current user?

File: src/messaging/views.py [hunk 7]
```diff
@@ -1,3 +1,157 @@
...
+
+    serializer_class = MessageSerializer
+
+    def get_queryset(self):
+        """
+        Filter messages to only show messages in communications where the user is a participant.
+        """
+        return Message.objects.filter(
+            communication__from_user=self.request.user
+        ) | Message.objects.filter(communication__to_user=self.request.user)
```
Comment (nomisreual) [inline src/messaging/views.py:22]: That does only filter for messages for which the current user is part of a communication. Looks like it's listing all messages though, not specific to one communication.

File: src/messaging/views.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nomisreual) [reply src/messaging/views.py:14]: see comment below

File: src/messaging/views.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nomisreual) [reply src/messaging/views.py:49]: looks good i think

File: src/messaging/views.py (hunk 1 above)
Comment (nomisreual) [reply src/messaging/views.py:106]: looks good.

File: src/messaging/urls.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+app_name = "messaging"
+
+urlpatterns = [
+    path("communications/", CommunicationList.as_view(), name="communication-list"),
+    path(
+        "communications/<int:pk>/",
+        CommunicationDetail.as_view(),
+        name="communication-detail",
+    ),
+    path("messages/", MessageList.as_view(), name="message-list"),
```
Comment (nomisreual) [inline src/messaging/urls.py:18]: maybe include the communication ID, so it is easier to fetch a list of messages related to one communication

**Review by nomisreual — CHANGES_REQUESTED** [review]

File: src/messaging/views.py (hunk 7 above)
Comment (ArezouRa) [reply src/messaging/views.py]: it's changed to get messages in a specific communication now :)

**Review by nomisreual — APPROVED** [review]
loioks good now

**Pull request merged**